### ISSUE TITLE
low mAP on 07test  trained on 07trainval

### ISSUE THREAD

**Issue by codegaj** [pierluigiferrari/ssd_keras#125]
First of all, thanks for your codes,  I want to reproduce the result of SSD300 on 07trainval dataset, after 60 epoches, I evaluated the model, just got mAP: 0.419
aeroplane     AP    0.411
bicycle       AP    0.553
bird          AP    0.231
boat          AP    0.191
bottle        AP    0.169
bus           AP    0.601
car           AP    0.572
cat           AP    0.679
chair         AP    0.147
cow           AP    0.295
diningtable   AP    0.404
dog           AP    0.561
horse         AP    0.66
motorbike     AP    0.604
person        AP    0.407
pottedplant   AP    0.184
sheep         AP    0.193
sofa          AP    0.496
train         AP    0.616
tvmonitor     AP    0.415
()
              mAP   0.419
here is the training csv:
epoch, training loss, val loss
![default](https://user-images.githubusercontent.com/17818478/40709018-886afc04-6427-11e8-80e6-22499838d396.png)

Something is wrong?

**Comment by pierluigiferrari** [comment]
If you train only on Pascal VOC 2007 `trainval`, of course you're going to get worse results than if you train on 2007 + 2012 `trainval`. 2012 `trainval` is more than twice as big as 2007 `trainval`, which means you're training on less than 30% of the data with which the benchmark results were achieved.

If you run [ssd300_training.ipynb](https://github.com/pierluigiferrari/ssd_keras/blob/master/ssd300_training.ipynb) as is, the mAP should be above 65% after 20,000 iterations. At least that's what happens every time I run it.

**Comment by codegaj** [comment]
thank you, I just want to reproduce results in the SSD paper, in SSD paper, the mAP trained on 07trainval is 68.0.

**Comment by pierluigiferrari** [comment]
Did you make any changes to the code or the parameters?

**Comment by codegaj** [comment]
no, I didn't,  and I found that the val loss is high.

**Comment by pierluigiferrari** [comment]
Two suggestions to better understand what's going on:
1. Run the training more than once, just to make sure that you get a similar result each time.
2. Run the training on 07+12 `trainval` for 20,000 steps to see if you get the expected result.

**Comment by codegaj** [comment]
OK, I will try. Thank you very mach for replying quickly.

**Comment by pierluigiferrari** [comment]
The second suggestion is probably the more important one to try first.

**Comment by pierluigiferrari** [comment]
I just stumbled upon the suspicious brackets in your original post above. Are you using Python 2?

**Comment by codegaj** [comment]
yes, I am using py2, but I change nothing in training process. In ssd300_evaluation.ipynb, the py3 package inspect has inspect.signature() function, but py2 doesn't have. I changed line 1071 in object_detection_2d_data_generator.py：
`if ('inverse_transform' in returns) and 'interpolation_mode' in dir(transform):#('return_inverter' in inspect.signature(transform).parameters):`
and line 1084:
`if ('inverse_transform' in returns) and 'interpolation_mode' in dir(transform): #('return_inverter' in inspect.signature(transform).parameters):`

I think the goal of signature() function is to identity whether the transfrom is Resize or ConvertTo3Channels, because only the Resize transform has parameter return_inverter. So, I changed like that, I think this change can also achieve the goal. Then , I test your ported weights: VGG_VOC0712_SSD_300x300_iter_120000.h5, and I get mAP 0.775, So, I think the change worked.

And, I just tried the second suggestion, unfortunately, after 20K steps, I got a mAP 0.416:
aeroplane     AP    0.497
bicycle       AP    0.564
bird          AP    0.298
boat          AP    0.166
bottle        AP    0.203
bus           AP    0.519
car           AP    0.47
cat           AP    0.712
chair         AP    0.217
cow           AP    0.21
diningtable   AP    0.299
dog           AP    0.629
horse         AP    0.617
motorbike     AP    0.535
person        AP    0.44
pottedplant   AP    0.159
sheep         AP    0.187
sofa          AP    0.448
train         AP    0.714
tvmonitor     AP    0.428
()
              mAP   0.416

And training loss, val loss:
![qq 20180531081234](https://user-images.githubusercontent.com/17818478/40756884-5122d460-64b8-11e8-8afe-1bacb154278d.png)


I feel so depressed, and I still try to solve the problem. Please don't mind my poor english, sorry.

**Comment by pierluigiferrari** [comment]
Then I suggest you create a Python 3 environment and try again. This implementation explicitly doesn't support Python 2.

I'm not super-familiar with all the differences between Python 2 and 3, but since you get an mAP of just 0.416 after 20k training steps on the original SSD300 "07+12" training and I consistently get an mAP of greater than 0.65 after 20k steps on the same training, our current best bet is that this difference might be caused by you using Python 2.

Just because the code works at all in Python 2 doesn't necessarily mean that it makes no difference whether you run it in Python 2 or Python 3. It may or may not make a difference. At least as long as something as basic as the Python version doesn't meet the requirements already, there is no point in trying to find the cause of the problem elsewhere.

**Comment by codegaj** [comment]
Thank you very much, bro. I think you can close the issue. I will try py3
